code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,489 @@
+        return
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(charge += EASY*T.on_impact(src, user))
+            true_target += list(process_target(user))
+            check_charge(user)
+
+/obj/item/xenoartifact/attack(atom/target, mob/user)
+    . = ..()
+    if(!(manage_cooldown(TRUE)))
+        return
+    var/intensity = istype(target, /mob/living) ? COMBAT : NORMAL
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(charge += intensity*T.on_impact(src, target))
+            true_target += list(process_target(target))
+            check_charge(user)      
+
+/obj/item/xenoartifact/afterattack(atom/target, mob/user, proximity)
+    . = ..()
+    if(!(manage_cooldown(TRUE))||proximity||get_dist(src, target) > max_range) //This proximity check might be considered messy, it's the result of various bugs. It works dont break it
+        return
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(charge += EASY*T.on_impact(src, user))
+            true_target += list(process_target(target))
+            check_charge(user)
+
+/obj/item/xenoartifact/throw_impact(atom/target, mob/user)
+    . = ..()
+    if(!(manage_cooldown(TRUE)))
+        return
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(charge += NORMAL*T.on_impact(src, user)) //Don't bother doing combat check here, thanks
+            true_target += list(process_target(target))
+            check_charge(user)
+
+/obj/item/xenoartifact/attackby(obj/item/I, mob/living/user)
+    for(var/datum/xenoartifact_trait/T in traits)
+        T.on_item(src, user, I)
+    if(!(manage_cooldown(TRUE))||user.a_intent == INTENT_HELP||istype(I, /obj/item/xenoartifact_label)||istype(I, /obj/item/xenoartifact_labeler))
+        return
+    var/impact_activator //Stinky approach
+    var/burn_activator
+    var/msg = I.ignition_effect(src, user)
+    for(var/datum/xenoartifact_trait/T in traits)
+        if(charge += NORMAL*T.on_impact(src, user, I.force, I))
+            impact_activator = TRUE
+        if(msg)
+            if(charge += NORMAL*T.on_burn(src, user, I.heat))
+                burn_activator = TRUE   
+                return    
+    if(impact_activator && !burn_activator)
+        true_target += list(process_target(user))
+        check_charge(user)
+    ..()
+/*
+    check_charge() is essentially what runs all the minor, major, and malf trait activations. 
+    This process also culls any irrelivent targets in reference to max_range and calculates the true charge.
+    True charge is simply, almost, the average of the charge and charge_req. This allows for a unique varience of 
+    output from artifacts, generally producing some funny results too.
+    
+*/
+/obj/item/xenoartifact/proc/check_charge(mob/user, charge_mod)
+    if(prob(malfunction_chance)) //See if we pick up an malfunction
+        var/datum/xenoartifact_trait/T = pick(subtypesof(/datum/xenoartifact_trait/malfunction))
+        traits[6] = new T
```

Comment:
What happens to the old trait when it gets overwritten? Should it be qdeleted or will byond GC handle that?

Reply:
From what I've seen, byond handles it. I could be wrong however.